musig/src/musig.rs

```diff
@@ -0,0 +1,312 @@
+    participants: Vec<PublicKey<E>>,
+    self_index: usize,
+    aggregated_public_key: PublicKey<E>,
+    a_self: E::Fs,
+
+    r_self: PrivateKey<E>,
+    r_pub_aggregated: PublicKey<E>,
+
+    t_others: Vec<Option<Vec<u8>>>,
+    t_self: Vec<u8>,
+    t_count: usize,
+
+    r_pub_others: Vec<Option<PublicKey<E>>>,
+    r_pub_self: PublicKey<E>,
+    r_pub_count: usize,
+
+    performed_sign: bool,
+}
+
+impl<E: JubjubEngine> MusigSession<E> {
+    pub fn new(
+        aggregate_hash: Box<dyn AggregateHash<E>>,
+        commitment_hash: Box<dyn CommitmentHash<E>>,
+        signature_hash: Box<dyn SignatureHash<E>>,
+        msg_hash: Box<dyn MsgHash>,
+        generator: FixedGenerators,
+        params: &E::Params,
+        participants: Vec<PublicKey<E>>,
+        seed: Seed<E>,
+        self_index: usize,
+    ) -> Result<Self, MusigError> {
+        let number_of_participants = participants.len();
+
+        if self_index >= number_of_participants {
+            return Err(MusigError::SelfIndexOutOfBounds);
+        }
+
+        let (aggregated_public_key, a_self) = MusigSession::<E>::compute_aggregated_public_key(
+            &participants,
+            &*aggregate_hash,
+            self_index,
+            params,
+        );
+
+        let (r_self, r_pub_self, t) =
+            MusigSession::<E>::generate_commitment(seed, &*commitment_hash, params, generator);
+
+        let session = MusigSession {
+            commitment_hash,
+            signature_hash,
+            msg_hash,
+            participants,
+            self_index,
+            aggregated_public_key,
+            a_self,
+            r_self,
+            r_pub_aggregated: r_pub_self.clone(),
+            t_others: vec![None; number_of_participants],
+            t_self: t,
+            t_count: 1,
+            r_pub_others: vec![None; number_of_participants],
+            r_pub_self,
+            r_pub_count: 1,
+            performed_sign: false,
+        };
+
+        Ok(session)
+    }
+
+    fn compute_aggregated_public_key(
+        participants: &[PublicKey<E>],
+        aggregate_hash: &dyn AggregateHash<E>,
+        self_index: usize,
+        params: &E::Params,
+    ) -> (PublicKey<E>, E::Fs) {
+        let mut x: Point<E, Unknown> = Point::zero();
+
+        let mut a_self = None;
+
+        for i in 0..participants.len() {
+            // TODO: Optimize
+            let ai = aggregate_hash.hash(&participants, &participants[i]);
+
+            x = x.add(&participants[i].0.mul(ai, params), params);
+
+            if i == self_index {
+                a_self = Some(ai);
+            }
+        }
+
+        let a_self = a_self.expect("Self index not in range");
+
+        (PublicKey(x), a_self)
+    }
+
+    fn generate_commitment(
+        seed: Seed<E>,
+        commitment_hash: &dyn CommitmentHash<E>,
+        params: &E::Params,
+        generator: FixedGenerators,
+    ) -> (PrivateKey<E>, PublicKey<E>, Vec<u8>) {
+        let r = PrivateKey::<E>(seed.0);
```

Comment:
Not all std::std can be a valid Path, so Path also provides some guarantees about what data it holds. When we're talking about Fs, PrivateKey and Seed they all hold exactly same data and the only difference is semantics and additional functions each of them has.
The author of original paper struggles himself to describe what this variable is, and he calls it just r and R. I just to decided use PrivateKey because:
1. It looks to me like ephemeral private key
2. We need R = r*G which also looks like ephemeral key pair and we have convenient function for that
3. It's unclear to me how changing type to Seed or Fs will help reader understand what is going on in this code

Reply:
Okay, I see your point. As a compromise I can suggest a solution which probably will help the reader more than the type name:

Let's add a comment explaining what's going on in this code. A couple of words about how `PrivateKey` is used to calculate `R` and what this `R` has to do with the signing process will reduce any kind of possible misunderstanding.

Is this OK for you?

Reply:
I think restoring algorithm from source code is not the best idea considering you have a paper that describes it. And if one read the paper, there's no need to describe what r is. From the other hand developers who doesn't want to dive into how this works can just follow readme and not think what r, t and other variables mean.

Reply:
Sure, then let it be as is.